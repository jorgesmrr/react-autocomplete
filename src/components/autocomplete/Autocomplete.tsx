import { MouseEventHandler, useEffect, useRef, useState } from "react";
import AutocompleteOptionComponent from "../autocomplete-option/AutocompleteOption";
import IconChevronDown from "../icons/IconChevronDown";
import IconClose from "../icons/IconClose";
import IconExclamationCircle from "../icons/IconExclamationCircle";
import Spinner from "../spinner/Spinner";
import styles from "./Autocomplete.module.css";

export interface AutocompleteOption {
  label: string;
  value: string | number;
}

type AutocompleteValue = string | number | null;

interface AutocompleteProps {
  value?: AutocompleteValue;
  onChange?: (value: AutocompleteValue) => void;
  typedValue: string;
  onChangeTypedValue: (value: string) => void;
  options: AutocompleteOption[] | null;
  id?: string;
  label?: string;
  tabIndex?: number;
  placeholder?: string;
  disabled?: boolean;
  isLoading?: boolean;
  emptyMessage?: string;
  errorMessage?: string;
  maxOptionsHeight?: string;
}

const Autocomplete: React.FC<AutocompleteProps> = ({
  value,
  onChange,
  typedValue,
  onChangeTypedValue,
  options,
  id,
  label,
  tabIndex,
  placeholder,
  disabled = false,
  isLoading = false,
  emptyMessage = "No suggestions for the typed text",
  errorMessage,
  maxOptionsHeight = "20rem",
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [isFocused, setFocused] = useState(false);
  const [highlightedOption, setHighlightedOption] = useState<number | null>(
    null
  );
  const [filteredOptions, setFilteredOptions] = useState<AutocompleteOption[]>(
    options || []
  );

  useEffect(() => {
    let newFilteredOptions: AutocompleteOption[];
    if (typedValue.length > 0) {
      newFilteredOptions = (options || []).filter(
        (option) =>
          option.label.toLowerCase().indexOf(typedValue.toLowerCase()) >= 0
      );
    } else {
      newFilteredOptions = options || [];
    }

    setFilteredOptions(newFilteredOptions);
    setHighlightedOption(newFilteredOptions.length ? 0 : null);
  }, [typedValue, options]);

  useEffect(() => {
    if (value !== null && value !== undefined) {
      const option = (options || []).find((option) => option.value === value);
      onChangeTypedValue(option ? option.label : "");
    } else {
      onChangeTypedValue("");
    }

    /* eslint react-hooks/exhaustive-deps: "off" */
  }, [value]);

  const selectOption = (option: AutocompleteOption | null) => {
    inputRef.current?.blur();
    onChange?.(option?.value || null);
    onChangeTypedValue(option?.label?.toString() || "");
  };

  const forceFocus = () => inputRef.current?.focus();

  const forceBlur = () => inputRef.current?.blur();

  const highlightPreviousOption = () => {
    setHighlightedOption(Math.max(0, (highlightedOption || 0) - 1));
  };

  const highlightNextOption = () => {
    setHighlightedOption(
      Math.min(filteredOptions.length - 1, (highlightedOption || 0) + 1)
    );
  };

  const selectHighlightedOption = () => {
    if (highlightedOption != null) {
      selectOption(filteredOptions[highlightedOption]);
    }
  };

  const handleClearClick = () => {
    selectOption(null);
    forceFocus();
  };

  const handleOptionMouseDown: MouseEventHandler<HTMLLIElement> = (event) => {
    // Prevent input blur when selecting an option
    event.preventDefault();
  };

  const handleOptionClick = (option: AutocompleteOption) => {
    selectOption(option);
  };

  // Key up event to not listen to multiple events in a single press
  const keyUpOperationsMap = {
    Escape: forceBlur,
    Enter: selectHighlightedOption,
  };

  // Key down event to listen to multiple events in a single press
  const keyDownOperationsMap = {
    ArrowUp: highlightPreviousOption,
    ArrowDown: highlightNextOption,
  };

  const handleKeyUp = (key: string) => {
    keyUpOperationsMap[key as keyof typeof keyUpOperationsMap]?.();
  };

  const handleKeyDown = (key: string) => {
    keyDownOperationsMap[key as keyof typeof keyDownOperationsMap]?.();
  };

  return (
    <div className={disabled ? styles.wrapperDisabled : styles.wrapper}>
      <div
        className={
          errorMessage ? styles.inputContainerDanger : styles.inputContainer
        }
      >
        {label && id && <label htmlFor={id}>{label}</label>}
        <input
          className={styles.input}
          ref={inputRef}
          id={id}
          type="text"
          tabIndex={tabIndex}
          placeholder={placeholder}
          disabled={disabled}
          value={typedValue}
          onChange={(e) => onChangeTypedValue(e.target.value)}
          onFocus={() => setFocused(true)}
          onBlur={() => setFocused(false)}
          onKeyUp={(e) => handleKeyUp(e.key)}
          onKeyDown={(e) => handleKeyDown(e.key)}
        />
        {typedValue.length > 0 && (
          <div className={styles.inputIcon} onClick={handleClearClick}>
            <IconClose />
          </div>
        )}
        {isLoading ? (
          <div className={styles.inputSpinner}>
            <Spinner />
          </div>
        ) : (
          <div className={styles.inputIcon} onClick={forceFocus}>
            <IconChevronDown />
          </div>
        )}
      </div>

      <div className={styles.popup}>
        {errorMessage && (
          <div className={styles.errorAlert}>
            <IconExclamationCircle />
            <span>{errorMessage}</span>
          </div>
        )}

        {isFocused && !isLoading && !errorMessage && (
          <ul className={styles.list} style={{ maxHeight: maxOptionsHeight }}>
            {typedValue.length > 0 && !filteredOptions.length ? (
              <li className={`${styles.listItemDisabled}`}>{emptyMessage}</li>
            ) : (
              filteredOptions.map((option, index) => (
                <li
                  key={option.value}
                  className={
                    highlightedOption === index
                      ? styles.listItemSelected
                      : styles.listItem
                  }
                  onMouseEnter={() => setHighlightedOption(index)}
                  onMouseDown={handleOptionMouseDown}
                  onClick={() => handleOptionClick(option)}
                >
                  <AutocompleteOptionComponent
                    option={option}
                    typedText={typedValue}
                  />
                </li>
              ))
            )}
          </ul>
        )}
      </div>
    </div>
  );
};

export default Autocomplete;
