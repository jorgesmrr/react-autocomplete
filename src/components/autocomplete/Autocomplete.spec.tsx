import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useState } from "react";
import Autocomplete from "./Autocomplete";

const options = [
  {
    label: "Dog",
    value: 1,
  },
  {
    label: "Cat",
    value: 2,
  },
  {
    label: "Mouse",
    value: 3,
  },
];

const WrapperComponent = () => {
  const [text, setText] = useState("");

  return (
    <Autocomplete
      typedValue={text}
      onChangeTypedValue={setText}
      options={options}
    />
  );
};

describe("Autocomplete", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<WrapperComponent />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should allow typing", async () => {
    const { getByRole } = render(<WrapperComponent />);

    const input = getByRole("textbox");

    const text = "Hey there!";
    userEvent.type(input, text);
    expect(input).toHaveValue(text);
  });

  it("should show items when the input is empty", async () => {
    const { getByRole, getAllByRole } = render(<WrapperComponent />);

    const input = getByRole("textbox");

    fireEvent.focus(input);
    expect(getAllByRole("listitem").length).toBe(options.length);
  });

  it("should select on pressing Enter", async () => {
    const { getByRole } = render(<WrapperComponent />);

    const input = getByRole("textbox");

    userEvent.type(input, "{enter}");
    expect(input).toHaveValue(options[0].label);
  });

  it("should correclty navigate with arrows", async () => {
    const { getByRole } = render(<WrapperComponent />);

    const input = getByRole("textbox");

    userEvent.type(input, "{arrowdown}{arrowdown}{arrowup}{enter}");
    expect(input).toHaveValue(options[1].label);
  });

  it("should hide suggestions on pressing Escape", async () => {
    const { getByRole, getAllByRole, queryByRole } = render(
      <WrapperComponent />
    );

    const input = getByRole("textbox");

    userEvent.type(input, "dog");
    expect(getAllByRole("listitem").length).toBeTruthy();

    userEvent.type(input, "{escape}");
    expect(queryByRole("listitem")).toBeNull();
  });
});
