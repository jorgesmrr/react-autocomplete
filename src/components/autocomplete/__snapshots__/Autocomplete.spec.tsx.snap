// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render correctly 1`] = `
<div
  className="wrapper"
>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="text"
      value=""
    />
    <div
      className="inputIcon"
      onClick={[Function]}
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </div>
  </div>
  <div
    className="popup"
  />
</div>
`;
